README.md

# RLE Zoo

[![Build status](https://github.com/eloj/rle-zoo/workflows/build/badge.svg)](https://github.com/eloj/rle-zoo/actions/workflows/c-cpp.yml)

A collection of Run-Length Encoders and Decoders, and associated tooling for exploring this space. So far there are only four animals in the zoo. It's a very small zoo.

* _WHILE THIS NOTE PERSISTS, I MAY FORCE PUSH TO MASTER_
* The codecs are written foremost to be robust, correct, and clear and easy to understand, not for performance.
* I'm mostly interested in exploring legacy formats.

## Features

* Single-Header Libraries.
* Releases are:
	* [Valgrind](https://valgrind.org/) clean,
	* [scan-build](https://clang-analyzer.llvm.org/scan-build.html) clean, and
	* [AFL++](https://aflplus.plus/) clean (for some reasonable run-time).

## What is Run-Length Encoding?

"_Run-length encoding (RLE) is a form of lossless data compression in which runs of data (sequences in which the same data value occurs in many consecutive data elements) are stored as a single data value and count, rather than as the original run._" -- [Wikipedia](https://en.wikipedia.org/wiki/Run-length_encoding)

At its most basic, a Run-Length Encoder process input into a series of REP (repeat) and CPY (copy) operations.
Sometimes the CPY operation is explicit, sometimes it's in the form of literals (LIT).

## Usage Example

The `rle_*.h` files are single-header libraries. If you just need one, any one, I recommend downloading `rle_packbits.h` and
looking at `test_example.c` for how to use it.

```c
#define RLE_ZOO_PACKBITS_IMPLEMENTATION
#include "rle_packbits.h"

int main(void) {
	const uint8_t input[] = "ABBBBA";
	size_t len = sizeof(input) - 1;

	// Call with NULL for dest buffer to calculate output size.
	ssize_t expected_size = packbits_compress(input, len, NULL, 0);
	...
```

## Tools

`rle-zoo` can encode and decode files using any of the supplied variants.

`rle-genops` can be used to generate complete code word/OPs lists for supported variants, and contains code that verifies
the encoding and decoding scheme for a variant is consistent. Post-implementation this is mostly useful for debugging,
'manual parsing' and reverse-engineering of unknown RLE streams. It can also generate C tables for implementing table-driven
encoders and decoders.

`rle-parser` can be used to parse a file using the available RLE variants, which could help identify the
variant used on some unknown data. It also acts as a demonstrator for using `rle-genops` tables. It
is a work in progress though, and _encoding is broken_ for some tables.

Example parsing a packbits encoded file:

```bash
$ hexdump -C tests/packbits/tn1023.rle
00000000  fe aa 02 80 00 2a fd aa  03 80 00 2a 22 f7 aa     |.....*.....*...|
0000000f
$ ./rle-parser -d -t packbits tests/packbits/tn1023.rle
Reading input from 'tests/packbits/tn1023.rle' (offset=0x0, max len=0x2000)
Parsing 15 byte buffer with 'packbits'
00000000: <fe> REP 3 'aa'
00000002: <02> CPY 3 ; 80 00 2a
00000006: <fd> REP 4 'aa'
00000008: <03> CPY 4 ; 80 00 2a 22
0000000d: <f7> REP 10 'aa'
Parse: rp=15, wp=24
```

Example parsing a file into packbits format:

```bash
$ hexdump -C tests/packbits/tn1023
00000000  aa aa aa 80 00 2a aa aa  aa aa 80 00 2a 22 aa aa  |.....*......*...|
00000010  aa aa aa aa aa aa aa aa                           |........|
$ ./rle-parser -e -t packbits tests/packbits/tn1023
Reading input from 'tests/packbits/tn1023' (offset=0x0, max len=0x2000)
Encoding 24 byte buffer with 'packbits'
Encode params = { cpy:{ 1, 128 }, rep:{2, 128} }
00000000: <fe> REP 3 'aa'
00000003: <02> CPY 3 ; 00 2a aa
00000006: <fd> REP 4 'aa'
0000000a: <03> CPY 4 ; 00 2a 22 aa
0000000e: <f7> REP 10 'aa'
rp=24, wp=15
```

## Zoo Animals

Currently the following extraordinary specimens are grazing the fertile grounds of this most amazing Zoo:

|  Variant  |  Type  | Code Use | REP Range | CPY/LIT Range | Spec. | Notes |
| :-------: | :----: | :------: | :-------: | :-----------: | :---: | :---- |
| [Packbits](#packbits) | CPY | Near-optimal | 2 - 128 | 1 - 128 | ref[^foottn1023] | One code (0x80) wasted on NOP. |
| [Goldbox](#goldbox) | CPY | Sub-optimal | 1 - 127 | 1 - 126 | n/a |  Used by [SSI Goldbox](https://en.wikipedia.org/wiki/Gold_Box) titles. Many quirks. |
| [PCX](#pcx) | LIT | Sub-optimal | 0 - 63 | 0 - 191 | [link](http://bespin.org/~qz/pc-gpe/pcx.txt) | |
| [Apple ICNS](#apple-icns) | CPY | Optimal | 3 - 130 | 1 - 128 | [ref](https://en.wikipedia.org/wiki/Apple_Icon_Image_format#Compression) | |

### PackBits

The `packbits` variant reportedly originates from the Apple Macintosh[^foottn1023], but spread far and wide from there, including
into Electronic Arts IFF ILBM (unverified).

#### PackBits Format

* One OP byte encoding the operation and `length`:
	* 0x00 => CPY 1
	* 0x01 => CPY 2
	* ..
	* 0x7e => CPY 127
	* 0x7f => CPY 128
	* 0x80 => _reserved_
	* 0x81 => REP 128
	* 0x82 => REP 127
	* ..
	* 0xfe => REP 3
	* 0xff => REP 2

* CPY: If OP high-bit is CLEAR (equally; signed OP is non-negative), then `OP + 1` literal bytes follows.
* REP: If OP high-bit is SET (equally; signed OP is negative), the next byte is repeated 257 - OP (or 1 - signed OP) times.

The encoder should not generate 0x80, which is reserved. The technical note from Apple states that a decoder should
"_[handle] this situation by skipping any flag-counter byte with this value and interpreting the next byte as the next flag-counter byte._"

The existance of the `NOP` and `REP 2` -- which requires two bytes to encode a run of two bytes -- are inefficiencies in the coding. Not terrible, but you can do better.

[^foottn1023]: "Understanding PackBits", Apple [Technical Note TN1023](http://web.archive.org/web/20080705155158/http://developer.apple.com/technotes/tn/tn1023.html).

### Goldbox

The `goldbox` variant has been organically hand-crafted to be 100% compatible with the classic Goldbox games _Pool of Radiance_ and _Curse of the Azure Bonds_,
and probably more. I have round-trip verified the decompressor and compressor on all resource files in both of those games, though the tests provided
in this repository are much more modest.

This variant is mostly interesting in the ways it's suboptimal; It's slightly short-stroked, and doesn't deal optimally with the end of the input.

#### Goldbox Format

* One OP byte encoding the operation and `length`:
	* 0x00 => CPY 1
	* 0x01 => CPY 2
	* ..
	* 0x7d => CPY 126
	* 0x7e => CPY 127 ; _not used by encoder_
	* 0x7f => CPY 128 ; _not used by encoder_
	* 0x80 => REP 128 ; _not used by encoder_
	* 0x81 => REP 127
	* 0x82 => REP 126
	* ..
	* 0xfe => REP 2
	* 0xff => REP 1
* CPY: If OP high-bit is CLEAR (equally; signed OP is non-negative), then `OP + 1` literal bytes follows.
* REP: If OP high-bit is SET (equally; signed OP is negative), then the next byte is to be repeated 1 + the bitwise complement of OP (or equally; the negation of the signed `OP`) times.

To the best of my knowledge, the encoder will never use OPs 0x7e, 0x7f or 0x80, i.e it will never output a literal run of 127 or 128 characters, nor a run of 128 repeated characters.

In addition to not using all the encodings, having both CPY 1 and REP 1 is obviously suboptimal.

It will also insist on encoding the tail of the input as a REP OP if it hits the end while counting literals to CPY.

To the last point, the input `1234` will be encoded as `02 31 32 33 ff 34`, i.e CPY 3 characters, then REP 1 character.

These limits and quirks were determined experimentally using existing game data files. The decoders in the games MAY accept
more optimal encodings, but the encoder provided here was specifically crafted such that encoding the output of the decoder
is bit-identitical to the original input.

### PCX

The `PCX` variant comes from the _ZSoft IBM PC Paintbrush_ software and its associated [PCX image format](https://en.wikipedia.org/wiki/PCX).
This image format was popular on the PC platform in the mid-1980s and well into the 1990s[^footpcxmame], but is now thoroughly obsolete.

The only compression ever defined for this format is a simple RLE variant which uses two bits of a byte to encode
REPs, and leaves the rest as literals.

[^footpcxmame]: The MAME project was using PCX up till May 1999; "Switched to PNG from PCX as the main screenshot image format", [Project Milestones](https://www.mamedev.org/history.html)

#### PCX Format

* One OP byte encoding REP and `length`, or used as-is:
	* 0x00 => LIT 0
	* 0x01 => LIT 1
	* ..
	* 0xbd => LIT 189
	* 0xbe => LIT 190
	* 0xbf => LIT 191
	* 0xc0 => REP 0
	* 0xc1 => REP 1
	* 0xc2 => REP 2
	* ..
	* 0xfe => REP 62
	* 0xff => REP 63
* REP: If the top two bits of OP are SET (equally; OP is >= 192), then the next byte is repeated `OP & 0x3F` times (0-63).
* LIT: Any other byte value is used as-is (0-191).

Obviously any single input byte larger than 191 MUST be encoded as a REP 1 -- expanding the output with one byte -- but an
encoder COULD encode any value using REP 1. In addition, depending on how the encoder is structured, it may encode two repeated
literals as-is, or as a REP 2. An encoder that scans for runs of literals will probably do the first, while one that output
literals as they are seen will probably do the former.

The existance of a `REP 0` operation is an inefficiency, and allows the encoder to encode data that is not included in the decoded output.

### Apple ICNS

Used by Apple to compress icon resources. This RLE variant uses a sensible encoding scheme, recognizing that
any `REP` with a count of 2 or less is redundant, and doesn't waste precious code space on a `NOP`.

#### Apple ICNS Format

* One OP byte encoding the operation and `length`:
	* 0x00 => CPY 1
	* 0x01 => CPY 2
	* ..
	* 0x7d => CPY 126
	* 0x7e => CPY 127
	* 0x7f => CPY 128
	* 0x80 => REP 3
	* 0x81 => REP 4
	* 0x82 => REP 5
	* ..
	* 0xfe => REP 129
	* 0xff => REP 130
* CPY: If high-bit is clear, then `OP + 1` literal bytes follow.
* REP: If high-bit is set, then repeat next byte `OP - 125` times.

The `CPY` OPs (0x00-0x7f) are the same as in packbits, but the `REP` OPs (0x80-) have been adjusted up to a minimum count of three.
They also come in ascending order compared with packbits; more characters are copied as the OP increase in value, versus fewer in packbits.

## TODO

* Add 'all' variant compression reporting to `rle-zoo`
* Make the rle-parse encoder follow limits/correct.
* Add more animals. Potential candidates: BMP(?), TGA, EXEPACK(?!), ...
* Improve `rle-zoo` to behave more like a standard UNIX filter.

## License

All code is provided under the [MIT License](LICENSE).


CHANGES.md
# RLE Zoo Headers & Tools

Stable releases have even minor version (e.g 1.0.x, 1.2.x, ..)

## Release 1.0.0-dev - next

* Headers and `rle-zoo` build MSVC CL v19.32.31332


rle-genops.c
/*
	Run-Length Encoding & Decoding Table Generator
	Copyright (c) 2022, Eddy L O Jansson. Licensed under The MIT License.

	See https://github.com/eloj/rle-zoo

	TODO:
		Use bitmap to mark off used RLE_OPS, then bit ops can detect
			missing or ambigious encodings.
		Output necessary types or include-file when --genc

*/
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <assert.h>
#include <string.h>
#include <limits.h>

#define UTILITY_IMPLEMENTATION
#include "utility.h"
#define RLE_PARSE_IMPLEMENTATION
#include "rle-parse.h"

#include "build_const.h"

typedef struct rle8 (*rle8_decode_fp)(uint8_t input);
typedef struct rle8 (*rle8_encode_fp)(struct rle8 cmd);

static int opt_usage, opt_genc;

static const char *gen_header = "// Generated by rle-genops from https://github.com/eloj/rle-zoo\n";

struct rle_parser {
	const char *name;
	rle8_encode_fp rle8_encode;
	rle8_decode_fp rle8_decode;
};

static struct rle8 rle8_decode_packbits(uint8_t input) {
	struct rle8 cmd;

	if (input > 0x80) {
		cmd.op = RLE_OP_REP;
		cmd.cnt = 1 - (int8_t)input;
	} else if (input < 0x80) {
		cmd.op = RLE_OP_CPY;
		cmd.cnt = input + 1;
	} else if (input == 0x80) {
		cmd.op = RLE_OP_NOP;
		cmd.cnt = 1;
	} else {
		// Impossible -- all decodings are valid
		cmd.op = RLE_OP_INVALID;
		cmd.cnt = 0;
	}

	assert((cmd.op != RLE_OP_REP) || (cmd.cnt <= 128 || cmd.cnt >= 2));
	assert((cmd.op != RLE_OP_CPY) || (cmd.cnt <= 128 || cmd.cnt >= 1));

	return cmd;
}

static struct rle8 rle8_encode_packbits(struct rle8 cmd) {
	struct rle8 res = { RLE_OP_INVALID, 0 };

	if (cmd.op == RLE_OP_REP) {
		if (cmd.cnt >= 2 && cmd.cnt <= 128) {
			res.op = RLE_OP_REP;
			res.cnt = 257 - cmd.cnt; // 1 - (int8_t)cmd.cnt;
		}
	} else if (cmd.op == RLE_OP_CPY) {
		if (cmd.cnt >= 1 && cmd.cnt <= 128) {
			res.op = RLE_OP_CPY;
			res.cnt = cmd.cnt - 1;
		}
	} else if (cmd.op == RLE_OP_NOP) {
		res.cnt = 0x80;
	}

	return res;
}

static struct rle8 rle8_decode_goldbox(uint8_t input) {
	struct rle8 cmd;

	if (input > 0x80) {
		cmd.op = RLE_OP_REP;
		cmd.cnt = (~input) + 1;
	} else if (input < 0x7e) {
		cmd.op = RLE_OP_CPY;
		cmd.cnt = input + 1;
	} else {
		cmd.op = RLE_OP_INVALID;
		cmd.cnt = 0;
	}

	assert((cmd.op != RLE_OP_REP) || (cmd.cnt <= 127 || cmd.cnt > 0));
	assert((cmd.op != RLE_OP_CPY) || (cmd.cnt <= 126 || cmd.cnt > 0));

	return cmd;
}

static struct rle8 rle8_encode_goldbox(struct rle8 cmd) {
	struct rle8 res = { RLE_OP_INVALID, 0 };

	if (cmd.op == RLE_OP_REP) {
		if (cmd.cnt >= 1 && cmd.cnt <= 127) {
			res.op = RLE_OP_REP;
			res.cnt = 1 + (~cmd.cnt);
		}
	} else if (cmd.op == RLE_OP_CPY) {
		if (cmd.cnt >= 1 && cmd.cnt <= 126) {
			res.op = RLE_OP_CPY;
			res.cnt = cmd.cnt - 1;
		}
	} // else if (cmd.op == RLE_OP_INVALID) res.cnt = 0x80; // XXX: Not sure why we'd want this. Add it as NOP in that case.

	return res;
}

static struct rle8 rle8_decode_pcx(uint8_t input) {
	struct rle8 cmd;

	if ((input & 0xC0) == 0xC0) {
		cmd.op = RLE_OP_REP;
		cmd.cnt = input & 0x3F;
	} else {
		cmd.op = RLE_OP_LIT;
		cmd.cnt = input;
	}

	assert((cmd.op != RLE_OP_REP) || (cmd.cnt <= 191 || cmd.cnt >= 0));
	assert((cmd.op != RLE_OP_LIT) || (cmd.cnt <= 63  || cmd.cnt >= 0));

	return cmd;
}

static struct rle8 rle8_encode_pcx(struct rle8 cmd) {
	struct rle8 res = { RLE_OP_INVALID, 0 };

	if (cmd.op == RLE_OP_REP) {
		if (cmd.cnt <= 63) {
			res.op = RLE_OP_REP;
			res.cnt = 0xC0 | cmd.cnt;
		}
	} else if (cmd.op == RLE_OP_LIT) {
		if (cmd.cnt <= 191) {
			res.op = RLE_OP_LIT;
			res.cnt = cmd.cnt;
		}
	}

	return res;
}

static struct rle8 rle8_decode_icns(uint8_t input) {
	struct rle8 cmd;

	if (input >= 0x80) {
		cmd.op = RLE_OP_REP;
		cmd.cnt = (int8_t)input - 125; // 3-130
	} else {
		cmd.op = RLE_OP_CPY;
		cmd.cnt = input + 1; // 1-128
	}

	assert((cmd.op != RLE_OP_REP) || (cmd.cnt <= 130 || cmd.cnt >= 3));
	assert((cmd.op != RLE_OP_CPY) || (cmd.cnt <= 128 || cmd.cnt >= 1));

	return cmd;
}

static struct rle8 rle8_encode_icns(struct rle8 cmd) {
	struct rle8 res = { RLE_OP_INVALID, 0 };

	if (cmd.op == RLE_OP_REP && (cmd.cnt >= 3 && cmd.cnt <= 130)) {
		res.op = RLE_OP_REP;
		res.cnt = cmd.cnt + 125;
	} else if (cmd.op == RLE_OP_CPY && (cmd.cnt >= 1 && cmd.cnt <= 128)) {
		res.op = RLE_OP_CPY;
		res.cnt = cmd.cnt - 1;
	}

	return res;
}

static int rle8_display_ops(struct rle_parser *p) {
	printf("// Automatically generated code table for RLE8 variant '%s'\n", p->name);
	printf("%s", gen_header);
	for (int i=0 ; i < 256 ; ++i) {
		uint8_t b = i;

		struct rle8 cmd = p->rle8_decode(b);

		if (cmd.op != RLE_OP_INVALID) {
			struct rle8 b_recode = p->rle8_encode(cmd);
			printf("0x%02x (%d/%d) => %s %d\n", b, b, (int8_t)b, rle_op_cstr(cmd.op), cmd.cnt);

			if (b != b_recode.cnt) {
				printf("ERROR: reencode mismatch: %s %d => 0x%02x\n", rle_op_cstr(cmd.op), cmd.cnt, b_recode.cnt);
				return 1;
			}
		} else {
			printf("0x%02x (%d/%d) => %s\n", b, b, (int8_t)b, rle_op_cstr(cmd.op));
		}
	}
	return 0;
}

static void rle8_generate_decode_table(struct rle_parser *p) {
	printf("\n// Decode table for RLE8 variant '%s'\n", p->name);

	printf("static struct rle8 rle8_tbl_decode_%s[256] = {\n", p->name);

	for (int i=0 ; i < 256 ; ++i) {
		uint8_t b = i;
		struct rle8 cmd = p->rle8_decode(b);
		printf(" /* %02X */ { RLE_OP_%s, %3d }", b, rle_op_cstr(cmd.op), cmd.cnt);
		if (i < 255) printf(",");
		if ((i < 255) && ((i+1) % 4) == 0) printf("\n");
	}

	printf("\n};\n");
}


static size_t rle8_generate_op_encode_array(struct rle_parser *p, int OP, char *buf, size_t bufsize) {
	// TODO: This should be autodetected by max of valid CPY/REP/LIT encoding, rest padded to -1
	int max_len = 256;

	size_t wp = 0;
	int tr = 0;
	buf_printf(buf, bufsize, &wp, &tr, "\t\t// RLE_OP_%s 0..%d\n", rle_op_cstr(OP), max_len - 1);
	buf_printf(buf, bufsize, &wp, &tr, "\t\t(int16_t[]){ ");

	for (int i=0 ; i < max_len ; ++i) {
		if (i > 0)
			buf_printf(buf, bufsize, &wp, &tr, ",");
		struct rle8 cmd = { OP, i };
		struct rle8 code = p->rle8_encode(cmd);
		if (code.op != RLE_OP_INVALID)
			buf_printf(buf, bufsize, &wp, &tr, "0x%02x", code.cnt);
		else
			buf_printf(buf, bufsize, &wp, &tr, "-1");
	}

	buf_printf(buf, bufsize, &wp, &tr, " },");

	if (tr) {
		fprintf(stderr, "INTERNAL ERROR: Buffer truncation detected -- output invalid!\n");
		exit(1);
	}

	return wp;
}

static void rle8_generate_encode_table(struct rle_parser *p) {
	// TODO: This should be autodetected by max of valid CPY/REP/LIT encoding, rest padded to -1
	// int max_len = 256;

	int minmax[3][2] = { { INT_MAX, INT_MIN }, { INT_MAX, INT_MIN }, { INT_MAX, INT_MIN } };
	int op_usage[RLE_OP_INVALID + 1] = { 0 };

	// Determine REP and CPY limits:
	for (int i=0 ; i < 256 ; ++i) {
		uint8_t b = i;
		struct rle8 cmd = p->rle8_decode(b);
		switch (cmd.op) {
			case RLE_OP_CPY:
			case RLE_OP_REP:
			case RLE_OP_LIT:
				if (cmd.cnt < minmax[cmd.op][0])
					minmax[cmd.op][0] = cmd.cnt;
				if (cmd.cnt > minmax[cmd.op][1])
					minmax[cmd.op][1] = cmd.cnt;
				break;
			case RLE_OP_NOP:
			case RLE_OP_INVALID:
				/* NOP */
				break;
		}
		op_usage[cmd.op]++;
	}

	char buf[1024+256] = {};
	size_t wp = 0;
	for (int i = RLE_OP_CPY ; i < RLE_OP_INVALID ; ++i) {
		if (op_usage[i] > 0) {
			buf_printf(buf, sizeof(buf), &wp, NULL, "%s(1U << RLE_OP_%s) /* %d */", wp == 0 ? "" : " | ", rle_op_cstr(i), op_usage[i]);
		}
	}

	printf("\nstatic struct rle8_tbl rle8_table_%s = {\n", p->name);
	printf("\t\"%s\",\n", p->name);
	printf("\t%s,\n", buf); // enum RLE_OP op_used;
	// SKIP FOR NOW printf("\t%d,\n", max_len);
	printf("\t{\n");
	for (int i = RLE_OP_CPY ; i < RLE_OP_NOP ; ++i) {
		if (op_usage[i] > 0) {
			rle8_generate_op_encode_array(p, i, buf, sizeof(buf));
			printf("%s\n", buf);
		} else {
			printf("\t\tNULL,\n");
		}
	}
	printf("\t},\n");
	printf("\trle8_tbl_decode_%s,\n", p->name);
	printf("\t{\n");
	for (int i = RLE_OP_CPY ; i < RLE_OP_NOP ; ++i) {
		if (op_usage[i] > 0) {
			printf("\t\t{ %d, %d }, // min-max %s\n", minmax[i][0], minmax[i][1], rle_op_cstr(i));
		} else {
			printf("\t\t{ -1, -1 }, // no %s\n", rle_op_cstr(i));
		}
	}
	printf("\t}\n};\n");

};

static void rle8_generate_c_tables(struct rle_parser *p) {
	printf("%s", gen_header);

	rle8_generate_decode_table(p);
	rle8_generate_encode_table(p);
}

struct rle_parser parsers[] = {
	{
		"goldbox",
		rle8_encode_goldbox,
		rle8_decode_goldbox
	},
	{
		"packbits",
		rle8_encode_packbits,
		rle8_decode_packbits
	},
	{
		"pcx",
		rle8_encode_pcx,
		rle8_decode_pcx
	},
	{
		"icns",
		rle8_encode_icns,
		rle8_decode_icns
	}
};
static const size_t NUM_VARIANTS = sizeof(parsers)/sizeof(parsers[0]);

static struct rle_parser* get_parser_by_name(const char *name) {
	for (size_t i = 0 ; i < NUM_VARIANTS ; ++i) {
		if (strcmp(name, parsers[i].name) == 0) {
			return &parsers[i];
		}
	}
	return NULL;
}

static void print_banner(void) {
	printf("rle-genops %s <%.*s>\n", build_version, 8, build_hash);
}

static void print_variants(void) {
	printf("\nAvailable variants:\n");
	struct rle_parser *p = parsers;
	for (size_t i = 0 ; i < NUM_VARIANTS ; ++i) {
		printf("  %s\n", p->name);
		++p;
	}
}

static void usage(const char *argv) {
	print_banner();
	printf("%s [OPTION] <variant>\n\n", argv);

	printf("Options:\n");
	printf("  --genc - Generate C tables.\n");

	print_variants();
}

static int parse_args(int *argc, char ***argv) {
	const char *arg = NULL;
	char **l_argv = *argv + 1;
	int l_argc = *argc - 1;
	int retval = 0;

	while ((arg = *l_argv) != NULL) {
		// "argv[argc] shall be a null pointer", section 5.1.2.2.1
		// const char *value = *l_argv;

		if (arg[0] == '-') {
			++l_argv;
			--l_argc;
			++arg;

			if (arg[0] == '-') {
				// long option
				++arg;
				if (strcmp(arg, "version") == 0) {
					print_banner();
					exit(0);
				} else if (strcmp(arg, "help") == 0) {
					opt_usage = 1;
				} else if (strcmp(arg, "genc") == 0) {
					opt_genc = 1;
				} else {
					retval = 1;
					break;
				}
			} else {
				// short option
				if (strcmp(arg, "v") == 0 || strcmp(arg, "V") == 0) {
					print_banner();
					exit(0);
				} else if (strcmp(arg, "h") == 0) {
					opt_usage = 1;
				} else {
					retval = 1;
					break;
				}
			}
		} else {
			break;
		}
	}

	if (retval != 0) {
		fprintf(stderr, "%s: unrecognized option: --%s\n", *argv[0], arg);
		fprintf(stderr, "Try '%s --help' for more information.\n", *argv[0]);
	}

	*argc = l_argc;
	*argv = l_argv;

	return retval;
}

int main(int argc, char *argv[]) {
	char **org_argv = argv;

	if (parse_args(&argc, &argv) != 0) {
		return 1;
	}

	if (opt_usage) {
		usage(org_argv[0]);
		return 1;
	}

	char *variant = argv[0];
	struct rle_parser *p = argc > 0 ? get_parser_by_name(variant) : NULL;

	if (!p) {
		if (variant) {
			fprintf(stderr, "error: Unknown variant '%s'\n", variant);
			print_variants();
		} else {
			usage(org_argv[0]);
		}
		return 2;
	}

	if (opt_genc)
		rle8_generate_c_tables(p);
	else
		rle8_display_ops(p);

	return EXIT_SUCCESS;
}
